### DISCUSSION TITLE
How to get the right time dimension of variable using series analysis wrapper?

### DISCUSSION THREAD

**Discussion by weiwilliam** [dtcenter/METplus#1143] [Configuration] [answered]
I would like to calculate the series analysis with the dataset below as an example using METplus 3.1 and MET 9.1 on Orion.
fcst files:
2015122100.nc
2015122106.nc
2015122112.nc
2015122118.nc
obs files:
20151221.nc (with 4 cycles available)

I've tried to setup METplus configuration file as below.
VALID_BEG=2015122100
VALID_END=2015122118
VALID_INCREMENT=6H
FCST_VAR1_LEVELS = "(0,0,*,*)"
OBS_VAR1_LEVELS = "({valid?fmt=%Y%m%d_%H%M%S},*,*)"
But it looks like {valid?fmt=%Y%m%d_%H%M%S} only use the VALID_BEG to generate the time dimension.
Is there any approach to make it find the right time dimension in obs dataset?

**Comment by georgemccabe** [comment]
Hi @weiwilliam,

If you were to upgrade to using METplus 4.0, then you would be able to get the correct valid times from this configuration. This version is available on Orion. Setup instructions can be found here https://dtcenter.org/community-code/metplus/metplus-4-0-existing-builds. You will need to set the following in your METplus config file:

`SERIES_ANALYSIS_RUNTIME_FREQ = RUN_ONCE_PER_INIT_OR_VALID`

This will run the tool once per valid time and substitute the correct valid time into OBS_VAR1_LEVELS for each run.

I recommend upgrading to v4.0, but if you must use METplus 3.1, please let me know. I believe you can configure the tool using that version to process these files by making a few changes including looping by initialization time, providing a forecast lead list, and modifying the filename templates.

**Reply by weiwilliam** [reply]
Thank you, @georgemccabe 
There is a problem with METplus 4.0 on Orion.
When I'm executing run_metplus.py, it will complains there is no netCDF available to import.
@jprestop suggested me to ask Orion helpdesk to install it.
I'll try your suggestion with METplus 4.0 when netCDF under intelpython3/2020.2 is ready.

About using METplus 3.1, do you mean to configure the file like below?
LOOP_BY = INIT
INIT_BEG=2015122100
INIT_END=2015122118
INIT_INCREMENT = 24H
LEAD_SEQ = 0, 6, 12, 18
FCST_VAR1_LEVELS = "(0,0,\*,\*)"
OBS_VAR1_LEVELS = "({init?fmt=%Y%m%d_%H%M%S},\*,\*)"

**Reply by georgemccabe** [reply]
The v3.1 configuration I was suggesting would look something like this:

```
LOOP_BY = INIT
INIT_BEG=2015122100
INIT_END=2015122100
INIT_INCREMENT = 24H
LEAD_SEQ = 0, 6, 12, 18
FCST_VAR1_LEVELS = "(0,0,*,*)"
OBS_VAR1_LEVELS = "({valid?fmt=%Y%m%d_%H%M%S},*,*)"

FCST_SERIES_ANALYSIS_INPUT_TEMPLATE = {valid?fmt=%Y%m%d%H}.nc
OBS_SERIES_ANALYSIS_INPUT_TEMPLATE = {valid?fmt=%Y%m%d}.nc
```

The SeriesAnalysis wrapper in 3.1 will only run once for a single INIT or VALID time (INIT_BEG or VALID_BEG), but it will process once for each forecast lead specified in LEAD_SEQ. This configuration will process the init time 2015-12-21 0Z but run once for each forecast lead to process valid times at 0Z, 6Z, 12Z, and 18Z. Since 4.0 the wrapper was enhanced to be able to be run for multiple init or valid times if desired, so this limitation is no longer present.

**Reply by weiwilliam** [reply]
Hi Geroge @georgemccabe ,

Thanks for your suggestion, it does get the proper valid time with METplus v4.0.
But I have further question,
When I try to get the statistics over these four cycles or a longer period, I set the file template following
```
FCST_SERIES_ANALYSIS_INPUT_TEMPLATE = {valid?fmt=%Y%m%d}*.nc
OBS_SERIES_ANALYSIS_INPUT_TEMPLATE = {valid?fmt=%Y%m%d}*.nc
```
It generates the filelist properly, but it complains messages below after the first cycle has been founded and matched.
ERROR  :
ERROR  : DataPlane::set_constant(double) -> no data buffer allocated!
ERROR  :
Is there any solution for this?
Thank you.

**Reply by georgemccabe** [reply]
Hi @weiwilliam, could you send the full log file for the run that produced this error?

**Reply by weiwilliam** [reply]
Here is it. Thanks.
[metplus.log.20210921190954.txt](https://github.com/dtcenter/METplus/files/7213105/metplus.log.20210921190954.txt)

**Reply by georgemccabe** [reply]
Thanks for posting this log file. A few things I noticed here:

1) It looks like the forecast file list only includes 1 file while the observation file list has 4.

> DEBUG 1: Length of forecast file list         = 1
> DEBUG 1: Length of observation file list      = 4

If you set LOG_LEVEL = DEBUG in your METplus user config file, the log file will contain more information about what files were put into the file list and could give some clues as to why it is not finding all of the files you need.

2) The current version of the SeriesAnalysis wrapper is designed to run once per field. In this case, you will need to be able to specify multiple levels with the time information in a single call so that the appropriate field is used for the observation file since it contains multiple fields (times) in the same file. An enhancement will have to be made to the wrapper to support this situation. I will create a GitHub issue to describe this enhancement. If you would like to provide the input files and the configuration files you are using (SeriesAnalysis.conf and SeriesAnalysisConfig_wrapped), I can use this case to test that the changes work as expected.

3) You are getting warnings saying that certain environment variables are not being utilized in your wrapped MET config file. This is OK because METplus wrappers 4.0.0 still supports the old variable names, however they may eventually become unsupported in future versions. This discussion page has resources for updating to use the wrapped config files that are in the METplus repository: https://github.com/dtcenter/METplus/discussions/996#discussioncomment-1304157
Also, please feel free to create a new GitHub Discussions topic if you want to make this update and need assistance.

**Reply by weiwilliam** [reply]
Hi George,
Sorry, I made a mistake with the log file.
The log file I uploaded is a test that I flipped the fcst and obs files templates.
So, METplus did find the files I need.
4 obs files are the fcst files at the beginning of this discussion thread. 
1 fcst file is the obs file.

I can create a folder on Orion for you to check, is it a good place for you?

Thank you.

**Reply by georgemccabe** [reply]
I don't have access to Orion, but I can have someone who does pull the files for me if that is easier for you. Another option is to upload the files to our FTP. Instructions on how to do that can be found here: #954

Here is the GitHub issue for this work: #1166
I will update this issue with more information once we come up with a good solution to handle this situation.

**Reply by weiwilliam** [reply]
Got it. I've created the tarball and uploaded it to FTP site.
It's test_data.tgz under Wei_data.
Under data folder, you'll find two subfolders. 
'fv3' keeps 4 fcst files and 'cams' keeps 1 obs file for the 4 cycles of 20151221
I also include the correct log file that found 4 fcst files and 1 obs file under 'logs'.
Thank you.

**Reply by georgemccabe** [reply]
It appears that there is something wrong with that tarfile. I get an error when I uncompress it and I only see the cams data, not the fv3 data. Could you regenerate the tarfile and upload it again?

**Reply by weiwilliam** [reply]
I uploaded the uncompressed folder. Please check if there is any problem, thanks.

**Reply by georgemccabe** [reply]
Thanks, I was able to get all of the data. Once the work is completed I will update this discussion with information on how to set up your config files.

**Comment by georgemccabe** [comment]
Hi @weiwilliam, I am going to mark this discussion as answered and lock it to prevent commenting. Please follow GitHub issue #1166 for this work for any updates related to the enhancements needed to fix this issue.

**Comment by georgemccabe** [comment, marked as answer]
Hi @weiwilliam , I am unlocking this discussion to give you an update. I was able to process your data with SeriesAnalysis. The changes required to get this to work are included in pull request #1353 and will be available in the METplus 4.1.0-beta5 release that will be created within a week. Here are a few of the settings that I changed from the config file you provided to get it to work:

```
SERIES_ANALYSIS_RUNTIME_FREQ = RUN_ONCE
FCST_VAR1_LEVELS = "(0,0,*,*)"
FCST_SERIES_ANALYSIS_INPUT_TEMPLATE = fv3_aods_{valid?fmt=%Y%m%d%H}_ll.nc
OBS_SERIES_ANALYSIS_INPUT_TEMPLATE = cams_aods_{valid?fmt=%Y%m%d}.nc
```

I also removed SERIES_ANALYSIS_CONFIG_FILE because you do not need to set this anymore. Leaving it unset will use the wrapped config file provided with the repository.

Please let me know if you have any other issues or questions.

**Reply by georgemccabe** [reply]
Hi @weiwilliam, I am just following up on this issue. Were you able to test to see if these changes work for your use case?

**Reply by weiwilliam** [reply]
Hi @georgemccabe, I haven't got a chance to test it.
Which platform has METplus 4.1.0-beta5?
I checked that is not available to load on Orion a few days ago.

**Reply by jprestop** [reply]
Hi @weiwilliam.

I have just installed METplus-4.1.0-beta5 on Orion.  I have updated our [Existing Builds](https://dtcenter.org/community-code/metplus/metplus-4-1-existing-builds) page to reflect this update.  

For quick reference, you can load the following to access this beta5 release:

> module use /apps/contrib/modulefiles
> module load metplus/4.1.0-beta5

Please let us know how it goes when you test the changes.  Thanks!

**Reply by weiwilliam** [reply]
Hi @georgemccabe and @jprestop , it works. Appreciate that you make this available. Thanks!
